packages/dds/shared-object-base/src/sharedObject.ts

```diff
@@ -519,3 +447,118 @@ export abstract class SharedObject<TEvent extends ISharedObjectEvents = ISharedO
 
     protected abstract applyStashedOp(content: any): unknown;
 }
+
+/**
+ * SharedObject with simplified, synchronous summarization and GC
+ */
+export abstract class SharedObject<TEvent extends ISharedObjectEvents = ISharedObjectEvents>
+    extends SharedObjectCore<TEvent> {
+    /**
+     * True while we are summarizing this object's data.
+     */
+    private _isSummarizing: boolean = false;
+
+    protected get serializer(): IFluidSerializer {
+        /**
+         * During summarize, the SummarySerializer keeps track of IFluidHandles that are serialized. These handles
+         * represent references to other Fluid objects and are used for garbage collection.
+         *
+         * This is fine for now. However, if we implement delay loading in DDss, they may load and de-serialize content
+         * in summarize. When that happens, they may incorrectly hit this assert and we will have to change this.
+         */
+        assert(!this._isSummarizing,
+            0x075 /* "SummarySerializer should be used for serializing data during summary." */);
+        return super.serializer;
+    }
+
+    /**
+     * @param id - The id of the shared object
+     * @param runtime - The IFluidDataStoreRuntime which contains the shared object
+     * @param attributes - Attributes of the shared object
+    */
+    constructor(
+        public id: string,
+        protected runtime: IFluidDataStoreRuntime,
+        public readonly attributes: IChannelAttributes)
+    {
+        super(id, runtime, attributes);
+    }
+
+    /**
+     * {@inheritDoc (ISharedObject:interface).summarize}
+     */
+    public summarize(fullTree: boolean = false, trackState: boolean = false): ISummaryTreeWithStats {
+        // Set _isSummarizing to true. This flag is used to ensure that we only use SummarySerializer (created below)
+        // to serialize handles in this object's data. The routes of these serialized handles are outbound routes
+        // to other Fluid objects.
+        assert(!this._isSummarizing, 0x076 /* "Possible re-entrancy! Summary should not already be in progress." */);
+        this._isSummarizing = true;
```

Comment:
I'm still not convinced that _isSummarizing (and certainly SummarySerializer) is useful during summarize rather than GC-only. Re-entrancy does seem like it shouldn't be happening, but I'm not sure that anything bad would result from it.

Reply:
So I believe the only purpose of this flag is to ensure that nobody uses this.serializer while collecting GC data, as that would miss some info (meaning local serializer would not be used).
I think you are right that this is not interesting for summarize() flow, only for GC collection flow.

And either way, I'd suggest removing it, and instead making this._serializer optional, and simply putting undefined there temporarily while collecting GC (and catching if anyone is touching it).

We can go even further and simply swapping this._serializer with another value while doing GC (and restoring it back one GC is over). I like this option way less as it's way more implicit and will not work for async flows.

Reply:
RE reentrancy: it mostly does not matter, with one caveat - if this._isSummarizing was a counter, not a boolean. Cause the primary purpose of this flag is not to catch reentrancy (it asserts there is no reentrancy because code dealing with this flag is wrong in presence of reentrancy), the purpose of this flag is to ensure nobody uses this.serializer while GC data is collected, that's it.

Reply:
I tried making serializer optional and putting undefined there while GCing. It doesn't seem any better to me, so I'd rather just keep _isSummarizing/_isGCing.